File: theorem-context-sdk/theorem-context-ts/src/adapters/claude_ai.ts
/**
 * Claude.ai conversation capture adapter (MEM-040).
 *
 * Sibling of `adapters/codex.ts`. Where the Codex adapter prepares a
 * Codex bundle from a Theorem context compile, this adapter goes the
 * other direction: it takes a captured claude.ai conversation array and
 * produces a session-summary payload that callers can feed to
 * `harness.memory.remember(...)`.
 *
 * The MD-file conversation
 * (`Theseus/Persistent-Interoperable agent memory ,Gap-Driven Fracture-Expansion, RustyRed-GraphBD.md`)
 * described a "/handoff claude-code" slash-command demo: capture the
 * claude.ai conversation, hand it off, spawn a Claude Code session
 * that boots with the same context. The realistic shape of that demo
 * is two-step:
 *
 *   1. `current_conversation(messages)` returns a session-summary
 *      payload built from the message array.
 *   2. `harness.memory.remember(payload)` saves it to the
 *      cross-surface memory store.
 *   3. `harness.action.handoff(...)` compiles a handoff artifact
 *      that the backend builds from the active workstream state
 *      (including the just-remembered summary).
 *
 * The host-side capture mechanism (claude.ai plugin, userscript, or
 * hook) is external to this repo; the adapter assumes a conversation
 * array is provided. Tests with fixture arrays verify the shaping
 * logic in isolation.
 */

export interface ConversationMessage {
  /** 'user' for human messages, 'assistant' for model replies. */
  role: 'user' | 'assistant';
  /** Raw text content of the message. */
  content: string;
  /** Optional ISO-8601 timestamp; used for first/last bookkeeping. */
  timestamp?: string;
}

export interface CapturedConversation {
  /**
   * Synthesized session-summary text. Compact, ~one-paragraph
   * description of the conversation's decisions and direction.
   * Suitable for `harness.memory.remember({observation})`.
   */
  observation: string;
  /**
   * Evidence pointers. Currently empty when the conversation is the
   * only source; callers can extend with URLs / commit hashes / file
   * paths the conversation referenced.
   */
  evidence: string[];
  /** Capture metadata for downstream provenance. */
  metadata: {
    surface: 'claude-ai';
    captured_at: string;
    message_count: number;
    first_message_at: string | null;
    last_message_at: string | null;
  };
}

/**
 * Build a session-summary payload from a claude.ai conversation array.
 *
 * The synthesis is intentionally simple: it concatenates message
 * content with role labels and adds a header that names the
 * conversation as a session_summary candidate. Callers who want
 * richer extraction (decisions, open questions, references) can
 * pass the raw `messages` array to a downstream LLM step before
 * calling `harness.memory.remember(...)`.
 */
export function currentConversation(
  messages: ConversationMessage[],
): CapturedConversation {
  const safeMessages = messages.filter(
    (msg) =>
      msg
      && typeof msg.content === 'string'
      && msg.content.trim().length > 0
      && (msg.role === 'user' || msg.role === 'assistant'),
  );

  const timestamps = safeMessages
    .map((msg) => (msg.timestamp || '').trim())
    .filter((t) => t.length > 0);

  const observationLines: string[] = ['Session summary captured from claude.ai conversation.'];
  for (const msg of safeMessages) {
    const role = msg.role === 'user' ? 'User' : 'Assistant';
    observationLines.push(`${role}: ${msg.content.trim()}`);
  }
  const observation = observationLines.join('\n\n');

  return {
    observation,
    evidence: [],
    metadata: {
      surface: 'claude-ai',
      captured_at: new Date().toISOString(),
      message_count: safeMessages.length,
      first_message_at: timestamps.length > 0 ? timestamps[0] : null,
      last_message_at: timestamps.length > 0 ? timestamps[timestamps.length - 1] : null,
    },
  };
}

/**
 * Alias matching the Python snake_case naming so cross-language docs
 * can use the same identifier. The canonical TS export is
 * `currentConversation`; `current_conversation` is an alias.
 */
export const current_conversation = currentConversation;

File: theorem-context-sdk/theorem-context-py/theorem_context/adapters/claude_ai.py
"""Claude.ai conversation capture adapter (MEM-041).

Python mirror of ``packages/theorem-context-ts/src/adapters/claude_ai.ts``.
Same shape, same synthesis logic. Pure function: no HTTP transport.

Usage::

    from theorem_context import (
        Harness,
        TheoremContextClient,
        current_conversation,
    )

    messages = [
        {'role': 'user', 'content': 'Should we ship Cluster D first?'},
        {'role': 'assistant', 'content': 'Yes, E depends on it.'},
    ]
    captured = current_conversation(messages)

    client = TheoremContextClient(api_key=...)
    harness = Harness(client=client)
    saved = await harness.memory.remember(
        captured['observation'],
        evidence=captured['evidence'],
    )
    handoff = await harness.action.handoff(
        workstream_id='ws-...',
        next_agent='claude-code',
    )

See ``docs/plans/memory-harness-mcp-post-fractal/handoff-demo.md`` for
the end-to-end developer demo (MEM-042).
"""

from __future__ import annotations

from datetime import datetime, timezone as _tz
from typing import Any


def current_conversation(
    messages: list[dict[str, Any]],
) -> dict[str, Any]:
    """Build a session-summary payload from a claude.ai conversation array.

    The synthesis is intentionally simple: it concatenates message
    content with role labels and adds a header that names the
    conversation as a session_summary candidate. Callers who want
    richer extraction (decisions, open questions, references) can
    run the raw ``messages`` through a downstream LLM step before
    calling ``harness.memory.remember(...)``.

    Returns a dict with keys ``observation``, ``evidence``, and
    ``metadata`` matching the TypeScript adapter's ``CapturedConversation``
    interface.
    """
    safe_messages = [
        msg
        for msg in (messages or [])
        if isinstance(msg, dict)
        and isinstance(msg.get('content'), str)
        and msg.get('content', '').strip()
        and msg.get('role') in ('user', 'assistant')
    ]

    timestamps = [
        (msg.get('timestamp') or '').strip()
        for msg in safe_messages
        if isinstance(msg.get('timestamp'), str)
        and (msg.get('timestamp') or '').strip()
    ]

    observation_lines = [
        'Session summary captured from claude.ai conversation.',
    ]
    for msg in safe_messages:
        role = 'User' if msg['role'] == 'user' else 'Assistant'
        observation_lines.append(f"{role}: {msg['content'].strip()}")
    observation = '\n\n'.join(observation_lines)

    return {
        'observation': observation,
        'evidence': [],
        'metadata': {
            'surface': 'claude-ai',
            'captured_at': datetime.now(_tz.utc).isoformat(),
            'message_count': len(safe_messages),
            'first_message_at': timestamps[0] if timestamps else None,
            'last_message_at': timestamps[-1] if timestamps else None,
        },
    }


__all__ = ['current_conversation']
